Holiday Snow: Add speed and hemisphere settings

tbradsha opened the pull request:
Closes MONOREP-271

> [!WARNING]
> This PR has been modified since it was first opened. 
> The hemisphere settings are no longer user facing ; they're automatically detected based on your timezone settings. See #46175 for more info. 

## Proposed changes:

This does a refactor of the class and adds speed ~and hemisphere settings to a dedicated settings page.~

In particular:

* Speed: A CSS var is added for the SCSS to use.
* Hemisphere: Now southern hemisphere users can get snow for the inverse limited season.

Other notes:

* I bumped the season through the 6th per this comment: p58i-3p3-p2#comment-30594 -> In some countries of the northern hemisphere, the holiday season lasts until the 6th (Reyes Magos in Spain is a good example)
* ~I'd love to avoid getting an option out-of-season, but couldn't think of a clean way to do that without repeated logic and still allowing the filter to work. Might have just been my EOD brain.~ This was fixed in #46175
* There are placeholders for some other potential settings (grid width, density), but at the moment they're not usable because SCSS uses the values pre-compilation.

Feel free to tag anyone else for review, disregard, close, or modify as desired!

Settings look like this:

### Other information:

- [x] Have you written new tests for your changes, if applicable?
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [x] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

## Jetpack product discussion
<!-- If you're an Automattician, include a shortlink to the p2 discussion with Jetpack Product here. -->
<!-- Make sure any changes to existing products have been discussed and agreed upon -->

## Does this pull request change what data or activity we track or use?

No

## Testing instructions:

* Go to Settings > General
* Ensure your site's timezone is set to a value using a city and not just a UTC offset. e.g. `Europe > London` or `Australia > Sydney`
    * If you've set it to a city in the northern hemisphere, you should see the Holiday section at the bottom of the General settings.
    * If you've set it to a city in the southern hemisphere, you should not see the Holiday section.
* Now change your site's timezone
* Check for the holiday snow section again.
* You can also use the `jetpack_is_holiday_snow_season` filter set to `true` to get things to show out of season.
* When snow can appear, try changing the snow settings and ensure the changes are reflected on the frontend.

tbradsha committed:
Refactor settings page and add some settings

tbradsha committed:
Use CSS var for speed

tbradsha committed:
Add some hemisphere settings

tbradsha committed:
Add changelogs

tbradsha committed:
Oops, we can only use config after we define it

tbradsha committed:
Phan tweaks

jeherve commented on projects/packages/jetpack-mu-wpcom/src/features/holiday-snow/class-holiday-snow.php:
```diff
@@ -15,21 +15,69 @@
+		} else {
+			$first_snow_day = mktime( 0, 0, 0, 12, 1 );
+			$last_snow_day  = mktime( 0, 0, 0, 1, 7 );
+		}
```
This seems problematic. I'm able to check the box, and snow will fall on my site today (December 3) even though it should only fall next summer for me:

<img width="1172" height="433" alt="Image" src="https://github.com/user-attachments/assets/f58a9a16-27a6-48dc-a12f-5e7a2e3bf049" />

I believe that's because June 1 and July 7 are within the same year, so the comparison below (`if ( $today >= $first_snow_day || $today < $last_snow_day ) {`) isn't correct for the southern hemisphere.

jeherve replied:
This was fixed in #46175

jeherve commented on projects/packages/jetpack-mu-wpcom/src/features/holiday-snow/class-holiday-snow.php:
```diff
@@ -76,14 +127,69 @@ public static function is_snow_enabled() {
+				'description' => __( 'If snow is enabled and this is ticked, it will show from 1 June through 6 July. Otherwise it will fall from 1 December through 6 January.', 'jetpack-mu-wpcom' ),
+				'label'       => __( 'Southern Hemisphere', 'jetpack-mu-wpcom' ),
+			),
+		);
```
Instead of making it an option, I was thinking it would be nice to have it detected automatically based on the site's timezone. Site owners would still be able to use a filter to set a custom snow season time, but we would do a better attempt at automatically figuring out the snow season for them.

This way we do not introduce a new setting that needs to be set by site owners, we fix the setting for them.

jeherve replied:
This behavior was changed in #46175

jeherve commented on projects/packages/jetpack-mu-wpcom/src/features/holiday-snow/class-holiday-snow.php:
```diff
@@ -186,49 +282,106 @@ public static function update_option_api( $input, $unfiltered_input ) {
+				continue;
+			}
+			register_setting(
+				self::SETTINGS_PAGE_SLUG,
```
I'm not sure we need a new settings page just for these 3 settings. It's a small feature, maybe it can continue to live under Settings > General?

jeherve replied:
I've removed the added settings page in #46175

jeherve committed:
Merge remote-tracking branch 'origin/trunk' into add/holiday-snow/settings

gmjuhasz commented on projects/packages/jetpack-mu-wpcom/src/features/holiday-snow/class-holiday-snow.php:
```diff
@@ -186,49 +282,106 @@ public static function update_option_api( $input, $unfiltered_input ) {
-				'sanitize_callback' => function ( $value ) {
-					return (bool) $value;
+		foreach ( self::$holiday_snow_config as $option_name => $option_config ) {
+			if ( $option_config['type'] === 'boolean' ) {
```
Nitpick: we could use Yoda contidion here as well as above in other places.

```suggestion
			if ( 'boolean' === $option_config['type'] ) {
```

jeherve replied:
👍 Done in 62cb10a724c027e75118d6047db58551e3772352

jeherve committed:
Holiday snow: switch to automatically detecting the hemisphere (#46175)

* Fix snow season calculation for southern hemisphere

* Holiday snow: switch to automatically detecting the hemisphere

Based on review here: https://github.com/Automattic/jetpack/pull/46139/files#r2584263969

Instead of asking the site owner to set an option for their site, let's try to make the decision for them, based on their site settings. If we know where their site is located (i.e. if their timezone includes a specific location), we can deduce their site's latitude, and thus the hemisphere they're in.

Since this commit removes that new field, I've opted to move the settings back to Settings > General, since I don't think we need a separate setting section for just 2 fields.

* Add changelog

* Add tests

* Fix test for PHP 8.5.

* Add more tests

* Be more defensive when fetching default config

jeherve committed:
Merge remote-tracking branch 'origin/trunk' into add/holiday-snow/settings

jeherve committed:
Switch to yoda conditions

See https://github.com/Automattic/jetpack/pull/46139#discussion_r2607397666

Co-authored-by: =?UTF-8?q?Gergely=20M=C3=A1rk=20Juh=C3=A1sz?=
 <36671565+gmjuhasz@users.noreply.github.com>

gmjuhasz reviewed (approved):
Works well for me!
